File: AalsiTrader/ViewModels/AdminViewModel.swift
import Foundation

@Observable
final class AdminViewModel {
    var users: [User] = []
    var stats: AdminService.SystemStats?
    var isLoading = false
    var error: String?
    var searchText = ""

    var filteredUsers: [User] {
        guard !searchText.isEmpty else { return users }
        return users.filter {
            $0.email.localizedCaseInsensitiveContains(searchText) ||
            $0.username.localizedCaseInsensitiveContains(searchText)
        }
    }

    func loadAll() async {
        isLoading = true
        error = nil
        do {
            async let u = AdminService.fetchUsers()
            async let s = AdminService.fetchStats()
            let (users, stats) = try await (u, s)
            self.users = users
            self.stats = stats
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func updatePlan(email: String, plan: PlanType, status: PlanStatus) async {
        do {
            try await AdminService.updatePlan(email: email, plan: plan, planStatus: status)
            await loadAll()
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func updateTrial(email: String, days: Int) async {
        do {
            try await AdminService.updateTrial(email: email, days: days)
            await loadAll()
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func updateTrading(email: String, enabled: Bool) async {
        do {
            try await AdminService.updateTrading(email: email, enabled: enabled)
            await loadAll()
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func updateAccount(email: String, enabled: Bool) async {
        do {
            try await AdminService.updateAccount(email: email, enabled: enabled)
            await loadAll()
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }

    func deleteUser(email: String) async {
        do {
            try await AdminService.deleteUser(email: email)
            users.removeAll { $0.email == email }
            HapticService.success()
        } catch {
            self.error = error.localizedDescription
            HapticService.error()
        }
    }
}

File: AalsiTraderAndroid/app/src/main/java/com/aalsitrader/android/viewmodel/AdminViewModel.kt
package com.aalsitrader.android.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aalsitrader.android.model.User
import com.aalsitrader.android.network.AdminStats
import com.aalsitrader.android.repository.AdminRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AdminViewModel : ViewModel() {
    private val repository = AdminRepository()

    private val _users = MutableStateFlow<List<User>>(emptyList())
    val users: StateFlow<List<User>> = _users.asStateFlow()

    private val _stats = MutableStateFlow<AdminStats?>(null)
    val stats: StateFlow<AdminStats?> = _stats.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _message = MutableStateFlow<String?>(null)
    val message: StateFlow<String?> = _message.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    init {
        loadAll()
    }

    fun loadAll() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try { _users.value = repository.getUsers() } catch (e: Exception) { _error.value = e.message }
            try { _stats.value = repository.getStats() } catch (_: Exception) { }
            _isLoading.value = false
        }
    }

    fun setSearchQuery(query: String) { _searchQuery.value = query }

    fun updatePlan(email: String, plan: String, status: String) {
        viewModelScope.launch {
            try {
                repository.updatePlan(email, plan, status)
                _message.value = "Plan updated"
                loadAll()
            } catch (e: Exception) { _error.value = e.message }
        }
    }

    fun updateTrial(email: String, days: Int) {
        viewModelScope.launch {
            try {
                repository.updateTrial(email, days)
                _message.value = "Trial updated"
                loadAll()
            } catch (e: Exception) { _error.value = e.message }
        }
    }

    fun toggleTrading(email: String, enabled: Boolean) {
        viewModelScope.launch {
            try {
                repository.updateTrading(email, enabled)
                _message.value = if (enabled) "Trading enabled" else "Trading disabled"
                loadAll()
            } catch (e: Exception) { _error.value = e.message }
        }
    }

    fun toggleAccount(email: String, enabled: Boolean) {
        viewModelScope.launch {
            try {
                repository.updateAccount(email, enabled)
                _message.value = if (enabled) "Account enabled" else "Account disabled"
                loadAll()
            } catch (e: Exception) { _error.value = e.message }
        }
    }

    fun clearError() { _error.value = null }
    fun clearMessage() { _message.value = null }

    fun filteredUsers(): List<User> {
        val query = _searchQuery.value
        return if (query.isBlank()) _users.value
        else _users.value.filter {
            it.email.contains(query, ignoreCase = true) ||
                    it.username.contains(query, ignoreCase = true)
        }
    }
}
